browser/ai_chat/ai_chat_tab_helper.cc

```diff
@@ -0,0 +1,300 @@
+  observers_.AddObserver(observer);
+}
+
+void AIChatTabHelper::RemoveObserver(Observer* observer) {
+  observers_.RemoveObserver(observer);
+}
+
+void AIChatTabHelper::RequestSummary() {
+  if (!article_summary_.empty()) {
+    OnArticleSummaryResult(article_summary_, true);
+    return;
+  }
+
+  ui::AXTreeID tree_id = web_contents()->GetPrimaryMainFrame()->GetAXTreeID();
+  content::RenderFrameHost* rfh =
+      content::RenderFrameHost::FromAXTreeID(tree_id);
+
+  rfh->GetMainFrame()->RequestAXTreeSnapshot(
+      base::BindOnce(&AIChatTabHelper::OnSnapshotFinished,
+                     base::Unretained(this)),
+      ui::AXMode::kWebContents, false, 5000, {});
+}
+
+void AIChatTabHelper::OnSnapshotFinished(const ui::AXTreeUpdate& snapshot) {
+  ui::AXTree tree;
+  if (!tree.Unserialize(snapshot)) {
+    return;
+  }
+
+  // Start AX distillation process
+  // Don't copy the tree, as it can be expensive.
+  DistillViaAlgorithm(tree);
+}
+
+void AIChatTabHelper::DistillViaAlgorithm(const ui::AXTree& tree) {
+  std::vector<const ui::AXNode*> content_root_nodes;
+  std::vector<const ui::AXNode*> content_nodes;
+  GetContentRootNodes(tree.root(), &content_root_nodes);
+
+  for (const ui::AXNode* content_root_node : content_root_nodes) {
+    AddContentNodesToVector(content_root_node, &content_nodes);
+  }
+
+  std::vector<std::u16string> text_node_contents;
+  for (const ui::AXNode* content_node : content_nodes) {
+    AddTextNodesToVector(content_node, &text_node_contents);
+  }
+
+  // TODO(nullhook): The assumption here is that 9300 chars equate to
+  // approximately 2k tokens, which is a rough estimate. A proper tokenizer is
+  // needed for accurate measurement.
+  auto contents_text =
+      base::JoinString(text_node_contents, u" ").substr(0, 9300);
+
+  VLOG(1) << __func__
+          << " Number of chars in content text = " << contents_text.length()
+          << "\n";
+
+  OnArticleSummaryResult(contents_text, false);
+}
+
+void AIChatTabHelper::OnArticleSummaryResult(const std::u16string& result,
+                                             bool is_from_cache) {
+  if (result.empty()) {
+    for (auto& obs : observers_) {
+      obs.OnRequestSummaryFailed();
+    }
+    return;
+  }
+
+  if (is_from_cache) {
+    VLOG(1) << __func__ << " Article summary is in cache\n";
+
+    AddToConversationHistory({
+        CharacterType::ASSISTANT,
+        ConversationTurnVisibility::VISIBLE,
+        base::UTF16ToUTF8(result),
+    });
+    return;
+  }
+
+  std::string summarize_prompt = base::ReplaceStringPlaceholders(
+      l10n_util::GetStringUTF8(IDS_AI_CHAT_SUMMARIZE_PROMPT),
+      {base::UTF16ToUTF8(result)}, nullptr);
+
+  // We hide the prompt with article content from the user
+  MakeAPIRequestWithConversationHistoryUpdate(
+      {CharacterType::HUMAN, ConversationTurnVisibility::HIDDEN,
+       summarize_prompt});
+}
+
+void AIChatTabHelper::CleanUp() {
+  chat_history_.clear();
+  article_summary_.clear();
```

Comment:
@sangwoo108  do I need to do any cleanups on the api helper?

Reply:
Well, you might want to clear `is_request_in_progress_` bit too and notify observers that. And possibly you could skip procedures in OnAPIResponse() when `is_request_in_progress_` is false?

Reply:
I have cleared is_request_in_progress_ for now. 

I think the cancellation of API requests should be notified to the API helper and properly cancel the request in its interface. This change is more involved, so I have added a TODO to resolve this as a follow-up.